Various post-merge fixes

nickelization opened the pull request:
This PR follows up on review comments from #105 with several changes to improve code quality:
- We now use application:get_env/3 instead of reimplementing its functionality ourselves
- Some test code in riak_ensemble_peer is now wrapped in `-ifdef(TEST)` macros
- The leadership watching code now uses monitors to detect dead watchers
- There is a new test module for verifying the leadership monitoring functionality

nickelization committed:
Use native get_env/3 instead of our own version

No reason to reimplement this functionality when we can just use the
code in the application module.

nickelization committed:
Only compile and export debug_local_get for tests

nickelization committed:
Switch leadership watcher cleanup to use monitors

This takes a bit of extra code, but is more idiomatic and
predictable than the old approach.

nickelization committed:
Add wait_until function to ens_test

This will come in handy for a new test I'm about to commit

nickelization committed:
Add get_watchers function for testing

This allows us to verify that watchers get removed from the internal
state when they unsubscribe or die.

nickelization committed:
Add leadership_watchers test

JeetKunDoug commented on src/riak_ensemble_peer.erl:
```diff
@@ -1850,12 +1863,17 @@ handle_event({watch_leader_status, Pid}, StateName, State) when node(Pid) =/= no
-    {next_state, StateName, State#state{watchers = NewWatcherList}};
+    case lists:keytake(Pid, 1, Watchers) of
+        false ->
+            lager:warning("Tried to stop watching for pid ~p, but did not find it in watcher list"),
```
Not sure if this shouldn't be a debug-level thing, rather than warning. In an operational sense, is anyone going to care if this happens in their cluster?

nickelization replied:
I was on the fence about this as well. My thinking was that a `stop_watching` for a pid that's not already watching indicates a bug in the system, so we want to make sure that such a scenario is noticed. (I even considered doing a pattern match that assumes we find the tuple, so that it crashes if it can't find it, but thought that would be a bit drastic since such a scenario doesn't necessarily indicate any crashworthy problems in the ensemble peer itself.)

From an operational standpoint, you're right, your typical sysadmin might not care about this. I was logging it as a warning from a more selfish perspective of wanting someone to notice it and say "hey Basho folk, I'm getting this message, should I worry?" Thus alerting us that there's a bug somewhere that should be fixed.

I can certainly appreciate the argument for making it a debug statement though. Thoughts?

nickelization replied:
Thought more about this, and I think my reasoning for making it a warning is not so good. Switched it to debug.

JeetKunDoug commented on src/riak_ensemble_peer.erl:
```diff
@@ -1884,22 +1902,29 @@ handle_sync_event(tree_info, _From, StateName, State=#state{tree_trust=Trust,
-    case Mod:handle_down(Ref, Pid, Reason, ModState) of
+handle_info({'DOWN', MRef, _, Pid, Reason}, StateName, State) ->
+    #state{mod=Mod, modstate=ModState, watchers=Watchers} = State,
+    case lists:keytake(Pid, 1, Watchers) of
```
not crazy about the nested case statements here - can we pull the `lists:keytake` part into a `remove_watcher` function that takes the pid and either returns a new list or `not_found` if it's not there? Also, in the `not_found` case I'd just call another function that encompasses the Mod:handle_down stuff - maybe `proxy_handle_down` or `mod_handle_down` or something to that effect? would make undertanding what's going on here easier I think.

I'd also add a quick comment explaining why there are two different ways we are dealing with `DOWN` messages.

nickelization replied:
Yeah funny you would suggest that, because I did also consider splitting the nested case out into separate functions. I thought it was still pretty understandable as is, so I decided to leave all the code in one place, but in hindsight I am probably biased since I was writing the code and not just reading it.

If it seemed obtuse to you then I'm happy to refactor it - will update the PR shortly.

nickelization committed:
Guard against the same pid watching twice

nickelization committed:
Switch leader status watch logging to debug

Seeing this message may be indicative of a bug, but most operators
aren't going to care about this particular message, so switch it from
warning to debug.

JeetKunDoug commented:
+1 8cb2b00

nickelization commented:
@borshop merge
